discovery/hetzner/hcloud.go

```diff
@@ -0,0 +1,122 @@
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package hetzner
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"net/http"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/go-kit/kit/log"
+	"github.com/hetznercloud/hcloud-go/hcloud"
+	config_util "github.com/prometheus/common/config"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/common/version"
+	"github.com/prometheus/prometheus/discovery/refresh"
+	"github.com/prometheus/prometheus/discovery/targetgroup"
+)
+
+const (
+	hetznerHcloudLabelPrefix                        = hetznerLabelPrefix + "hcloud_"
+	hetznerLabelHcloudImageName                     = hetznerHcloudLabelPrefix + "image_name"
+	hetznerLabelHcloudImageDescription              = hetznerHcloudLabelPrefix + "image_description"
+	hetznerLabelHcloudPrivateIPv4                   = hetznerHcloudLabelPrefix + "private_ipv4"
+	hetznerLabelHcloudDatacenterLocation            = hetznerHcloudLabelPrefix + "datacenter_location"
+	hetznerLabelHcloudDatacenterLocationNetworkZone = hetznerHcloudLabelPrefix + "datacenter_location_network_zone"
+	hetznerLabelHcloudCPUCores                      = hetznerHcloudLabelPrefix + "cpu_cores"
+	hetznerLabelHcloudCPUType                       = hetznerHcloudLabelPrefix + "cpu_type"
+	hetznerLabelHcloudMemoryGB                      = hetznerHcloudLabelPrefix + "memory_size_gb"
+	hetznerLabelHcloudDiskGB                        = hetznerHcloudLabelPrefix + "disk_size_gb"
+	hetznerLabelHcloudType                          = hetznerHcloudLabelPrefix + "server_type"
+)
+
+// Discovery periodically performs Hetzner Cloud requests. It implements
+// the Discoverer interface.
+type hcloudDiscovery struct {
+	*refresh.Discovery
+	client *hcloud.Client
+	port   int
+}
+
+// newHcloudDiscovery returns a new hcloudDiscovery which periodically refreshes its targets.
+func newHcloudDiscovery(conf *SDConfig, logger log.Logger) (*hcloudDiscovery, error) {
+	d := &hcloudDiscovery{
+		port: conf.Port,
+	}
+
+	rt, err := config_util.NewRoundTripperFromConfig(conf.HTTPClientConfig, "hetzner_sd", false)
+	if err != nil {
+		return nil, err
+	}
+	d.client = hcloud.NewClient(
+		hcloud.WithApplication("Prometheus", version.Version),
+		hcloud.WithHTTPClient(&http.Client{
+			Transport: rt,
+			Timeout:   time.Duration(conf.RefreshInterval),
+		}),
+		hcloud.WithEndpoint(conf.hcloudEndpoint),
+	)
+	return d, nil
+}
+func (d *hcloudDiscovery) refresh(ctx context.Context) ([]*targetgroup.Group, error) {
+	servers, err := d.client.Server.All(ctx)
+	if err != nil {
+		return nil, err
+	}
+	networks, err := d.client.Network.All(ctx)
```

Comment:
The API seems self contained, is this needed?

Reply:
The servers endpoint just answers with the network ID (https://docs.hetzner.cloud/#servers-get-a-server) so we need to get the list of all networks first